File: lib/ingestion/pipeline.ts
// Document ingestion pipeline. Port of legacy/ingestion/pipeline.py.
//
// Flow per file:
//   1. Drop raw file in Storage at <customer>/<package_id>/original.<ext>
//   2. Extract Markdown via lib/ingestion/extract (Claude vision for PDF/img)
//   3. Save Markdown alongside the raw file
//   4. Classify into one of the document categories
//   5. Save metadata.json
//   6. Append a DOCUMENT_INGESTED event with the package id + category
//
// Drive mirror is intentionally deferred — Supabase Storage is canonical now.

import crypto from "crypto";
import { appendEvent } from "@/lib/events/events";
import { extract } from "@/lib/ingestion/extract";
import { classifyDocument, type DocumentCategory } from "@/lib/ingestion/classifier";
import { uploadFile, ensureBucket } from "@/lib/ingestion/storage";

export interface IngestFile {
  filename: string;
  content: Buffer;
  mimeType: string;
  source: "slack" | "email" | "upload" | "drive" | "unknown";
  sourceDetail?: string;
  metadata?: Record<string, unknown>;
}

export interface IngestResult {
  packageId: string;
  category: DocumentCategory;
  pageCount: number;
  markdownPath: string;
  originalPath: string;
  metadataPath: string;
  status: string;
}

export function slugify(s: string, max = 60): string {
  return (
    s
      .toLowerCase()
      .replace(/\.[^.]+$/, "")
      .replace(/[^\w\-]+/g, "-")
      .replace(/^-+|-+$/g, "")
      .slice(0, max) || "untitled"
  );
}

export async function ingest(customerKey: string, file: IngestFile): Promise<IngestResult> {
  await ensureBucket();
  const ts = new Date();
  const md5 = crypto.createHash("md5").update(file.content).digest("hex").slice(0, 8);
  const slug = slugify(file.filename);
  const stamp = ts.toISOString().replace(/[-:]/g, "").replace(/\..+$/, "").replace("T", "");
  const packageId = `${customerKey}-${slug}_${stamp}_${md5}`;
  const ext = (file.filename.match(/\.([a-zA-Z0-9]+)$/)?.[1] ?? "bin").toLowerCase();

  // 1 + 3 — store original and markdown side by side.
  const originalPath = await uploadFile(
    {
      customerKey,
      packageId,
      filename: file.filename,
      content: file.content,
      contentType: file.mimeType,
    },
    `original.${ext}`
  );

  // 2 — extract markdown.
  const { markdown, pageCount } = await extract(file.content, file.mimeType);

  const markdownPath = await uploadFile(
    {
      customerKey,
      packageId,
      filename: `${slug}.md`,
      content: markdown,
      contentType: "text/markdown; charset=utf-8",
    },
    "content.md"
  );

  // 4 — classify.
  const category = await classifyDocument(file.filename, markdown);

  // Mirror into a category-keyed copy so the agent's search and the dashboard
  // organised view both have a stable lookup path.
  const orgPath = await uploadFile(
    {
      customerKey,
      packageId,
      filename: `${slug}.md`,
      content: markdown,
      contentType: "text/markdown; charset=utf-8",
    },
    `organized/${category}/${slug}.md`
  );
  const orgOrigPath = await uploadFile(
    {
      customerKey,
      packageId,
      filename: file.filename,
      content: file.content,
      contentType: file.mimeType,
    },
    `original_docs/${category}/${file.filename}`
  );

  const metadata = {
    id: packageId,
    title: file.filename,
    source: file.source,
    source_detail: file.sourceDetail ?? "",
    ingested_at: ts.toISOString(),
    original_filename: file.filename,
    mime_type: file.mimeType,
    md5,
    page_count: pageCount,
    category,
    organized_path: orgPath,
    original_doc_path: orgOrigPath,
    status: "indexed",
    extra: file.metadata ?? {},
  };

  const metadataPath = await uploadFile(
    {
      customerKey,
      packageId,
      filename: "metadata.json",
      content: JSON.stringify(metadata, null, 2),
      contentType: "application/json",
    },
    "metadata.json"
  );

  // 6 — log event so the agent and dashboard see the new doc.
  await appendEvent(
    customerKey,
    "DOCUMENT_INGESTED",
    {
      filename: file.filename,
      package_id: packageId,
      category,
      source: file.source,
      mime_type: file.mimeType,
      page_count: pageCount,
      organized_path: orgPath,
      original_doc_path: orgOrigPath,
    },
    {
      summary: `Ingested ${file.filename} → ${category}/`,
      tags: ["document", category, file.source],
    }
  );

  return {
    packageId,
    category,
    pageCount,
    markdownPath,
    originalPath,
    metadataPath,
    status: `Ingested ${file.filename} → ${category}/ (${pageCount} page${pageCount === 1 ? "" : "s"}).`,
  };
}

File: legacy/ingestion/pipeline.py
"""Ingestion pipeline — orchestrates: detect type → convert → index → organize → log."""

from __future__ import annotations

import hashlib
import json
import logging
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from slugify import slugify

from curator.customers import get_customer
from curator.ingestion import converters, classifier
from curator.storage import gdrive, event_log, cache_sync

logger = logging.getLogger(__name__)


# ── File abstraction ──────────────────────────────────────────────────────────


@dataclass
class IngestFile:
    """Represents a file to be ingested, regardless of source."""

    filename: str
    content: bytes
    mime_type: str = "application/octet-stream"
    source: str = "unknown"  # "slack", "email", "upload"
    source_detail: str = ""  # e.g. channel name, email subject
    metadata: dict[str, Any] = field(default_factory=dict)


# ── Pipeline ──────────────────────────────────────────────────────────────────


def _compute_md5(data: bytes) -> str:
    return hashlib.md5(data).hexdigest()


async def ingest(customer_key: str, file: IngestFile) -> str:
    """Full ingestion: inbox → indexed_documents → organized_docs → event log → cache sync.

    Returns a human-readable status string.
    """
    from curator.customers import ensure_customer_drive_folder
    customer = get_customer(customer_key)
    drive_folder_id = await ensure_customer_drive_folder(customer_key)
    timestamp = datetime.now(timezone.utc)
    md5_short = _compute_md5(file.content)[:8]
    slug = slugify(file.filename, max_length=60)
    package_id = f"{customer_key}-{slug}_{timestamp:%Y%m%d%H%M%S}_{md5_short}"

    # Determine file extension
    ext = file.filename.rsplit(".", 1)[-1] if "." in file.filename else "bin"

    logger.info("[%s] Ingesting %s (%s, %d bytes)", customer_key, file.filename, file.mime_type, len(file.content))

    # ── 1. Drop original in /inbox/ ───────────────────────────────────────
    inbox_name = f"{timestamp.isoformat()}_{file.filename}"
    await gdrive.upload(drive_folder_id, f"inbox/{inbox_name}", file.content, file.mime_type)

    # ── 2. Create indexed_documents package ───────────────────────────────
    package_path = f"indexed_documents/{package_id}"
    await gdrive.create_folder(drive_folder_id, package_path)
    await gdrive.upload(drive_folder_id, f"{package_path}/original.{ext}", file.content, file.mime_type)

    # ── 3. Convert based on type ──────────────────────────────────────────
    pages: list[bytes] = []
    page_map: list[dict[str, Any]] = []
    markdown = ""

    match file.mime_type:
        case "application/pdf":
            pages, markdown, page_map = await converters.pdf_to_package(file.content)
            for i, page_img in enumerate(pages):
                await gdrive.upload(
                    drive_folder_id,
                    f"{package_path}/pages/page_{i + 1:03d}.png",
                    page_img,
                    "image/png",
                )
            if page_map:
                await gdrive.upload(
                    drive_folder_id,
                    f"{package_path}/text/page_map.json",
                    json.dumps(page_map),
                )

        case "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
            markdown = await converters.docx_to_markdown(file.content)

        case "application/vnd.openxmlformats-officedocument.presentationml.presentation":
            markdown = await converters.pptx_to_markdown(file.content)

        case "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet":
            markdown = await converters.xlsx_to_markdown(file.content)

        case t if t.startswith("image/"):
            markdown = await converters.image_to_markdown(file.content)

        case _:
            markdown = await converters.generic_to_text(file.content)

    # Upload the markdown content
    await gdrive.upload(drive_folder_id, f"{package_path}/text/content.md", markdown)

    # ── 4. Write metadata.json ────────────────────────────────────────────
    metadata = {
        "id": package_id,
        "title": file.filename,
        "source": file.source,
        "source_detail": file.source_detail,
        "ingested_at": timestamp.isoformat(),
        "original_filename": file.filename,
        "mime_type": file.mime_type,
        "md5": md5_short,
        "page_count": len(pages) if pages else None,
        "status": "indexed",
    }

    # ── 5. Classify and place in organized_docs + original_docs ─────────
    category = await classifier.classify(markdown, file.filename)
    metadata["category"] = category
    org_path = f"organized_docs/{category}/{slug}.md"
    metadata["organized_path"] = org_path
    await gdrive.upload(drive_folder_id, org_path, markdown)

    # Store the original (unprocessed) document in original_docs
    orig_doc_path = f"original_docs/{category}/{file.filename}"
    metadata["original_doc_path"] = orig_doc_path
    await gdrive.upload(drive_folder_id, orig_doc_path, file.content, file.mime_type)

    await gdrive.upload(
        drive_folder_id,
        f"{package_path}/metadata.json",
        json.dumps(metadata, indent=2),
    )

    # ── 6. Log event ─────────────────────────────────────────────────────
    await event_log.append(
        customer_key,
        "DOCUMENT_INGESTED",
        {
            "filename": file.filename,
            "package_id": package_id,
            "category": category,
            "source": file.source,
            "mime_type": file.mime_type,
            "page_count": len(pages) if pages else None,
        },
        summary=f"Ingested {file.filename} → {category}/",
    )

    # ── 7. Sync to cache ─────────────────────────────────────────────────
    try:
        await cache_sync.sync_customer(customer_key)
    except Exception:
        logger.warning("Post-ingest cache sync failed for %s", customer_key, exc_info=True)

    status = f"📄 Ingested {file.filename} → {category}/ ({len(pages) if pages else '?'} pages)"
    logger.info("[%s] %s", customer_key, status)
    return status
